{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "commercial-typing",
   "metadata": {},
   "source": [
    "# Single-cell Analysis Workflow: 1.3M brain cells example\n",
    "**Author:** [Severin Dicks](https://github.com/Intron7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda0ac25-cdbc-451f-84a9-d56a65fec2c0",
   "metadata": {},
   "source": [
    "To run this notebook please make sure you have a working RAPIDS environment with all necessary dependencies. In this example workflow we'll be looking at a dataset of 1.3M brain cells from [10X Genomics](https://www.10xgenomics.com/datasets/1-3-million-brain-cells-from-e-18-mice-2-standard-1-3-0)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fabulous-ultimate",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import cupy as cp\n",
    "\n",
    "import time\n",
    "import rapids_singlecell as rsc\n",
    "\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "twenty-orbit",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rmm\n",
    "from rmm.allocators.cupy import rmm_cupy_allocator\n",
    "\n",
    "rmm.reinitialize(\n",
    "    managed_memory=False,  # Allows oversubscription\n",
    "    pool_allocator=False,  # default is False\n",
    "    devices=0,  # GPU device IDs to register. By default registers only GPU 0.\n",
    ")\n",
    "cp.cuda.set_allocator(rmm_cupy_allocator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6957ee36-066d-4904-ad13-7d22785fa119",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "scenic-overview",
   "metadata": {},
   "source": [
    "## Load and Prepare Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06463480-029b-4ea6-9bad-aaac52d736ed",
   "metadata": {},
   "source": [
    "Let's start by ensuring that we have our dataset downloaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "89406d0e-ecd2-4911-a406-a2a1fa8404a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./h5 directory found\n",
      "./h5/nvidia_1.3M.h5ad dataset found\n"
     ]
    }
   ],
   "source": [
    "import wget\n",
    "import os\n",
    "from anndata.experimental import read_elem_lazy\n",
    "import h5py\n",
    "\n",
    "url = 'https://exampledata.scverse.org/rapids-singlecell/1M_brain_cells_10X.sparse.h5ad'\n",
    "data_dir = \"./h5\"\n",
    "final = data_dir+'/nvidia_1.3M.h5ad'\n",
    "\n",
    "if not os.path.exists(data_dir): # Check if h5 directory exists\n",
    "    print('creating data directory')\n",
    "    os.system('mkdir ./h5')\n",
    "else:\n",
    "    print(f'{data_dir} directory found')\n",
    "\n",
    "if not os.path.isfile(final): # Check to see if we have our final output.  If it is there, get to the analysis!\n",
    "    print(f'Downloading cell data into {final}...')\n",
    "    wget.download(url, final)\n",
    "else:\n",
    "    print(f'{final} dataset found')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "saved-plain",
   "metadata": {},
   "source": [
    "We load the sparse count matrix from an `h5ad` file using Scanpy. The sparse count matrix will then be placed on the GPU. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "original-election",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_load_start = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "comic-fundamental",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adata = sc.read(final)\n",
    "adata.var_names_make_unique()\n",
    "adata = adata[:1_000_000, :].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "assisted-blind",
   "metadata": {},
   "source": [
    "We now load the the AnnData object into VRAM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e89dabcc-691c-488c-882f-bcaa65078425",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.get.anndata_to_GPU(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55ea4758-0763-48bb-a77f-864f27319842",
   "metadata": {},
   "source": [
    "Verify the shape of the resulting sparse matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73990ba5-0426-436a-9552-3e710c957033",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8ca7e2d-a04c-4b9c-ba7b-b1518fb942f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_load_time = time.time()\n",
    "print(\"Total data load and format time: %s\" % (data_load_time - data_load_start))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e5267d2-f8bd-433f-b56e-c349cd0eef25",
   "metadata": {},
   "source": [
    "## Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd732eb6-95b3-49e1-b6e7-14178dcb9b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "preprocess_start = time.time()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dcd1582-c0b8-43ce-a2ab-c02888a23364",
   "metadata": {},
   "source": [
    "### Quality Control"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd63a058-cdcc-44a2-9cbe-a595e6810b90",
   "metadata": {},
   "source": [
    "We perform a basic qulitiy control and plot the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ede69ff-c59e-4fd4-9cd6-823fbbc15355",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.flag_gene_family(adata, gene_family_name=\"MT\", gene_family_prefix=\"mt-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69b88a61-40f0-4156-8a16-eacc625cfa94",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.calculate_qc_metrics(adata, qc_vars=[\"MT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a063d03-a86c-4697-a7cf-10d358d839b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pl.scatter(adata, \"total_counts\", \"pct_counts_MT\")\n",
    "sc.pl.scatter(adata, \"total_counts\", \"n_genes_by_counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "993700c1-531b-4be5-8700-fb335714368b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pl.violin(adata, keys=\"n_genes_by_counts\")\n",
    "sc.pl.violin(adata, keys=\"total_counts\")\n",
    "sc.pl.violin(adata, keys=\"pct_counts_MT\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05c6f920-5ee2-4f85-a4b7-73f8f6c7294c",
   "metadata": {},
   "source": [
    "### Filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b167fa3-494f-445c-a61b-dbe356c7e1b9",
   "metadata": {},
   "source": [
    "We filter the count matrix to remove cells with an extreme number of genes expressed.\n",
    "We also filter out cells with a mitchondrial countent of more than 20%."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11798efe-23e8-4aba-8899-887a7c77d017",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adata = adata[\n",
    "    (adata.obs[\"n_genes_by_counts\"] < 5000)\n",
    "    & (adata.obs[\"n_genes_by_counts\"] > 500)\n",
    "    & (adata.obs[\"pct_counts_MT\"] < 20)\n",
    "].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8eb60cb9-1a6b-4194-baff-35d021954cde",
   "metadata": {},
   "source": [
    "Many python objects are not deallocated until garbage collection runs. When working with data that barely fits in memory (generally, >50%) you may need to manually trigger garbage collection to reclaim memory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8404fcc6-f598-44b0-a993-1ae2502a0b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fefb279a-bc73-4c10-a192-1c73e2fbf53a",
   "metadata": {},
   "source": [
    "We also filter out genes that are expressed in less than 3 cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b01fc2a5-acd1-411c-976c-dd689dfa289f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.filter_genes(adata, min_cells=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5cd75bc-623b-4944-b5ec-3f2c61dd5833",
   "metadata": {},
   "source": [
    "We store the raw expression counts in the `.layer[\"counts\"]`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e6a239c-23b6-432b-8c0f-3e934d69aa06",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers[\"counts\"] = adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc12e1d4-8529-4171-99ea-8d98080be042",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80db10de-f3f9-4f47-91e3-9129d2f7a4bb",
   "metadata": {},
   "source": [
    "### Normalize"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7720a81e-fa2e-4161-bb3d-3b9d74a417a8",
   "metadata": {},
   "source": [
    "We normalize the count matrix so that the total counts in each cell sum to 1e4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31f6017d-6564-4cde-87fa-b21a59689716",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.normalize_total(adata, target_sum=1e4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed79bbb4-4ca8-4fd8-8244-36858faae96d",
   "metadata": {},
   "source": [
    "Next, we log transform the count matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f04851b4-41fc-4226-966e-58985dcf4432",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbc151a6-9c2a-414c-b17d-8923afa47310",
   "metadata": {},
   "source": [
    "### Select Most Variable Genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a3c271-d114-4e9d-95b4-bf119f1b947b",
   "metadata": {},
   "source": [
    "Now we search for highly variable genes. This function only supports the flavors `cell_ranger` `seurat` `seurat_v3` and `pearson_residuals`. As you can in scanpy you can filter based on cutoffs or select the top n cells. You can also use a `batch_key` to reduce batcheffects."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7a8a962-cebf-4877-80ef-b1bf8ec5a5ee",
   "metadata": {},
   "source": [
    "In this example we use `seurat_v3` for selecting highly variable genes based on the raw counts in `.layer[\"counts\"]`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ba8a4e7-3d91-4b0b-b3c1-0fba00e41b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.highly_variable_genes(\n",
    "    adata, n_top_genes=5000, flavor=\"seurat_v3\", layer=\"counts\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d42cd76-81af-4c99-b9f1-318576507e54",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.get.anndata_to_CPU(adata, layer=\"counts\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96c3d84b-a950-4a75-a303-dbbedafe4b40",
   "metadata": {},
   "source": [
    "Now we safe this version of the AnnData as adata.raw."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5fc8d19-2765-4e37-9645-d2a2912b2550",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e66acd3c-1e98-4880-a8c1-620325c5903c",
   "metadata": {},
   "source": [
    "Now we restrict our AnnData object to the highly variable genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99becef5-b23f-479a-86aa-20641f6e25ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adata = adata[:, adata.var[\"highly_variable\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2eeaba9-00a4-4062-8b1d-bb18168a4b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f8f3372-ac66-4704-bfa7-8b0ec685eec7",
   "metadata": {},
   "source": [
    "Next we regress out effects of counts per cell and the mitochondrial content of the cells. As you can with scanpy you can use every numerical column in `.obs` for this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24ae8802-9134-486b-962d-b65f48e9de72",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.regress_out(adata, keys=[\"total_counts\", \"pct_counts_MT\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b589fa68-1ee5-4f7a-b34d-208ab6c08484",
   "metadata": {},
   "source": [
    "### Scale"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45c20a8d-6017-4325-8d3a-ae9ea69253c0",
   "metadata": {},
   "source": [
    "Finally, we scale the count matrix to obtain a z-score and apply a cutoff value of 10 standard deviations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c55cdec7-9efa-41c9-8955-c1c4389fbf4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.scale(adata, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf95e30a-362d-4357-bcf6-4a572b9d4037",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Principal component analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2de9aacf-69ad-4f2f-b7bc-5ac5685f0cca",
   "metadata": {},
   "source": [
    "We use PCA to reduce the dimensionality of the matrix to its top 100 principal components. We use the PCA implementation from cuml to run this. With `use_highly_variable = False` we save VRAM since we already subset the matrix to only HVGs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4d61564-b4c6-4ff3-bcab-be538db91c26",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.pca(adata, n_comps=100, use_highly_variable=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdbfed87-872c-438a-a6e2-e5bfe9cc1359",
   "metadata": {},
   "source": [
    "We can use scanpy `pca_variance_ratio` plot to inspect the contribution of single PCs to the total variance in the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35f516b0-e017-417d-9bb4-7879ef785e54",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca_variance_ratio(adata, log=True, n_pcs=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42508420-3aff-4a56-8b70-ff3ac579e798",
   "metadata": {},
   "source": [
    "Now we move `.X` and `.layers` out of the GPU."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c9728c5-0390-4d0d-8273-27c948bf0dcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.get.anndata_to_CPU(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66c90635-ff33-40b2-a9ba-4b991c1b295e",
   "metadata": {},
   "outputs": [],
   "source": [
    "preprocess_time = time.time()\n",
    "print(\"Total Preprocessing time: %s\" % (preprocess_time - preprocess_start))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "first-reggae",
   "metadata": {},
   "source": [
    "Visualization## Clustering and Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "imported-meeting",
   "metadata": {},
   "source": [
    "### Computing the neighborhood graph and UMAP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "biological-phenomenon",
   "metadata": {},
   "source": [
    "Next we compute the neighborhood graph using rsc.\n",
    "\n",
    "Scanpy CPU implementation of nearest neighbor uses an approximation, while the GPU version calculates the exact graph. Both methods are valid, but you might see differences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "transparent-major",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.pp.neighbors(adata, n_neighbors=15, n_pcs=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "photographic-script",
   "metadata": {},
   "source": [
    "Next we calculate the UMAP embedding using rapdis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "timely-actor",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.tl.umap(adata, min_dist=0.3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "linear-highlight",
   "metadata": {},
   "source": [
    "### Clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wooden-submission",
   "metadata": {
    "tags": []
   },
   "source": [
    "Next, we use the Louvain and Leiden algorithm for graph-based clustering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "certified-mixer",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.tl.louvain(adata, resolution=0.6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "initial-ribbon",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.tl.leiden(adata, resolution=1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "frozen-convertible",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pl.umap(adata, color=[\"louvain\", \"leiden\"], legend_loc=\"on data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee072829-de51-49c8-857e-24f89b9b7ec0",
   "metadata": {},
   "source": [
    "## TSNE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79004b0e-da2c-459d-8127-03df131b8810",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.tl.tsne(adata, n_pcs=40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45507d64-15f8-4f0c-9753-5227abb9e9ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata, color=[\"louvain\", \"leiden\"], legend_loc=\"on data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sexual-delaware",
   "metadata": {},
   "source": [
    "## Differential expression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fallen-exposure",
   "metadata": {},
   "source": [
    "We now use logistic regression to compute a ranking for highly differential genes in each Louvaincluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "naked-treasury",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.tl.rank_genes_groups_logreg(adata, groupby=\"louvain\", use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "national-grenada",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "sc.pl.rank_genes_groups(adata, n_genes=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "verbal-programming",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Diffusion Maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c3a493e-b018-4bac-ae87-f856fd868bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "rsc.tl.diffmap(adata)\n",
    "adata.obsm[\"X_diffmap\"] = adata.obsm[\"X_diffmap\"][:, 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affiliated-excess",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.diffmap(adata, color=\"louvain\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "informational-dealer",
   "metadata": {},
   "source": [
    "After this you can use `X_diffmap` for `sc.pp.neighbors` and other functions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ranking-brazilian",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "print(\"Total Processing time: %s\" % (time.time() - preprocess_start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44134fc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These notebooks are very GPU memory intensive!\n",
    "# In order to free up GPU memory, we'll kill this kernel prior to proceeding.  You will get a message.  This is expected.\n",
    "# If you have a CUDA or an Out Of Memory (OOM) error, please kill all kernels to free up your GPU memory and try again!\n",
    "# You can comment this out if you want to continue exploring the notebook.\n",
    "# Please consult the README for more tips and tricks.\n",
    "\n",
    "import IPython\n",
    "\n",
    "IPython.Application.instance().kernel.do_shutdown(True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
